libs/utils/src/sync/heavier_once_cell.rs

```diff
@@ -455,4 +476,94 @@ mod tests {
+        init_take_deinit_scenario(|cell, factory| {
+            Box::pin(async {
+                cell.get_or_init(factory).await.unwrap();
+            })
+        })
+        .await;
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn reproduce_init_take_deinit_race_mut() {
+        init_take_deinit_scenario(|cell, factory| {
+            Box::pin(async {
+                cell.get_mut_or_init(factory).await.unwrap();
+            })
+        })
+        .await;
+    }
+
+    type BoxedInitFuture<T, E> = Pin<Box<dyn Future<Output = Result<(T, InitPermit), E>>>>;
+    type BoxedInitFunction<T, E> = Box<dyn Fn(InitPermit) -> BoxedInitFuture<T, E>>;
+
+    /// Reproduce an assertion failure with both initialization methods.
+    ///
+    /// This has interesting generics to be generic between `get_or_init` and `get_mut_or_init`.
+    /// Alternative would be a macro_rules! but that is the last resort.
+    async fn init_take_deinit_scenario<F>(init_way: F)
+    where
+        F: for<'a> Fn(
+            &'a OnceCell<&'static str>,
+            BoxedInitFunction<&'static str, Infallible>,
+        ) -> Pin<Box<dyn Future<Output = ()> + 'a>>,
+    {
+        let cell = OnceCell::default();
+
+        // acquire the init_semaphore only permit to drive initializing tasks in order to waiting
+        // on the same semaphore.
+        let permit = cell
+            .inner
+            .read()
+            .await
+            .init_semaphore
+            .clone()
+            .try_acquire_owned()
+            .unwrap();
+
+        let mut t1 = pin!(init_way(
+            &cell,
+            Box::new(|permit| Box::pin(async move { Ok(("t1", permit)) })),
+        ));
+
+        let mut t2 = pin!(init_way(
+            &cell,
+            Box::new(|permit| Box::pin(async move { Ok(("t2", permit)) })),
+        ));
+
+        // drive t2 first to the init_semaphore
+        tokio::select! {
+            _ = &mut t2 => unreachable!("it cannot get permit"),
+            _ = tokio::time::sleep(Duration::from_secs(3600 * 24 * 7 * 365)) => {}
+        }
+
+        // followed by t1 in the init_semaphore
+        tokio::select! {
+            _ = &mut t1 => unreachable!("it cannot get permit"),
+            _ = tokio::time::sleep(Duration::from_secs(3600 * 24 * 7 * 365)) => {}
+        }
+
+        // now let t2 proceed and initialize
+        drop(permit);
+        t2.await;
+
+        let (s, permit) = { cell.get_mut().await.unwrap().take_and_deinit() };
+        assert_eq!("t2", s);
+
+        // now originally t1 would see the semaphore it has as closed. it cannot yet get a permit from
+        // the new one.
+        tokio::select! {
+            _ = &mut t1 => unreachable!("it cannot get permit"),
+            _ = tokio::time::sleep(Duration::from_secs(3600 * 24 * 7 * 365)) => {}
+        }
+
+        // only now we get to initialize it
+        drop(permit);
+        t1.await;
+
+        assert_eq!("t1", *cell.get().await.unwrap());
+    }
```

Comment:
@problame:

> Not sure if the init_take_deinit_scenario test cases are worth it. I would have preferred a failpoint-style approach. Obviously not with failpoint's global state, but, a cfg(test) field in the heavier_once_cell to configure the failpoint.

I don't understand your comment about failpointing. How would that make this easier to test than the semaphore queue? This test is fully deterministic because of that. The complication in this test comes from running the same test for both `get_or_init` or `get_mut_or_init`.

Reply:
I stared another 10min at that test and still don't get what's happening :/

Reply:
So there's the internal semaphore, also note that `tokio::test(start_paused = true)` is being used. `start_paused = true` when used together with `t1` and `t2` runs them to `Poll::Pending` and then advances time by a year, releasing the `sleep`.

In the main body we first acquire the only permit from within the `OnceCell`, then in turn first run `t2` into the semaphores inner queue by running it for a "year". Then repeat for `t1`. At that point, the next permit the semaphore will receive will go to `t2` and after that to `t1`. They are both stuck on line 85: `sem.acquire().await`, but *will* only be awoken from there in order of `t2`, `t1`.

Now on test line 550 we drop the "testing permit" which allows `t2` to complete the `sem.acquire().await` and complete the whole initialization.

On the line 553 we are keeping `t1` suspended or "unlucky" and `take_and_deinit` as we discussed while `t1` is waiting on the **old semaphore**. Because this PR version does not require `take_and_deinit` take all permits (unlike the follow-up), we complete right away.

On line 558 we run `t1` to acquire a permit from the **new semaphore** which line 553 switched. It cannot get it, because we are holding it (`permit` on line 553).

On line 563 we drop the `permit` from 553 and so `t1` gets to complete, and finally assert that `t1` did complete.

Reply:
> and then advances time by a year, releasing the sleep.

I saw the `start_paused` but didn't realize it's [this tokio feature.](https://tokio.rs/tokio/topics/testing#pausing-and-resuming-time-in-tests)

Specifically, this aspect wasn't clear to me _at all_:

> Pausing time has the effect that any time-related future may become ready early. The condition for the time-related future resolving early is that there are no more other futures which may become ready.

I mean, this is neat, but, I think it warrants some comment hint.

---

With that established, the test makes a lot more sense to me.